framework/docs/source/how-to-upgrade-to-message-api.rst

```diff
@@ -0,0 +1,384 @@
+   ``Message``.
+
+.. note::
+
+    The payload that `Message` objects carry is of type |recorddict_link|_ which can
+    contain records of type ``ArrayRecord``, ``MetricRecord`` and ``ConfigRecord``.
+
+.. code-block:: python
+    :emphasize-lines: 9,10,18,23,33,34,37,38,46,56,57
+
+    from flwr.client import ClientApp
+    from flwr.common import ArrayRecord, Context, Message, MetricRecord, RecordDict
+    from my_utils import train_fn, test_fn
+
+    # Flower ClientApp
+    app = ClientApp()
+
+
+    @app.train()
+    def train(msg: Message, context: Context) -> Message:
+        """Train the model on local data."""
+
+        # Init Model and data loader
+        train_loader = DataLoader(...)
+        model = MyModel()
+
+        # Read ArrayRecord received from ServerApp
+        arrays = msg.content["arrays"]
+        # Load weights to model
+        model.load_state_dict(arrays.to_torch_state_dict())
+
+        # Do local training
+        train_loss = train_fn(model, train_loader)
+
+        # Construct reply Message: arrays and metrics
+        model_record = ArrayRecord(model.state_dict())
+        # You can include any metric (scalar or list of scalars)
+        # relevant to your usecase.
+        # A weighting metric (`num-examples` by default) is always
+        # expected by FedAvg to do aggregation
+        metrics = MetricRecord(
+            {
+                "train_loss": train_loss,
+                "num-examples": len(train_loader.dataset),
+            }
+        )
+        # Construct RecorDict and add ArrayRecord and MetricRecord
+        content = RecordDict({"arrays": model_record, "metrics": metrics})
+        return Message(content=content, reply_to=msg)
+
+
+    @app.evaluate()
+    def evaluate(msg: Message, context: Context) -> Message:
+        """Evaluate the model on local data."""
+
+        # Identical to @app.train but returning only metrics
+        # after doing local evaluation
+        # ...
+
+        # Do local evaluation
+        loss, accuracy = test_fn(model, test_loader)
+
+        # Construct reply Message
+        # Retrun metrics relevant to usecase
+        # THe weighting metric is also sent and will be used
+        # to do weighted aggregation of metrics
+        metrics = MetricRecord(
+            {
+                "eval_loss": loss,
+                "eval_accuracy": accuracy,
+                "num-examples": len(test_loader.dataset),
+            }
+        )
+        # Construct RecorDict and add MetricRecord
+        content = RecordDict({"metrics": metrics})
+        return Message(content=content, reply_to=msg)
+
+This concludes the migration guide!
+
+.. tip::
+
+    If you would like to create a new Flower App using the `Message API`, run the ``flwr
+    new`` command and choose the appropriate template. Alternatively, you may want to
+    take a look at the `quickstart-pytorch
+    <https://github.com/adap/flower/blob/main/examples/quickstart-pytorch>`_ example.
```

Comment:
Let's move this tip to the top. It'll be easier for users to migrate their existing apps if they have a fully working projects as a starting point.